File: WHO all groups/Cuba 2/Setup_model.m
clear all;

states1     = {'U'};
states2     = {'Lf','Ls','Pf','Ps','I','I2','Tx','Tx2','Rlo','Rhi','R'};
gps.age     = {'ch','ad'};                                                  % added age here before other groups
gps.born    = {'dom','vuln'};
gps.strains = {'ds'};
gps.hiv     = {'neg','pos','art'};

[i, s, d, lim] = get_addresses({states1, gps.age, gps.born, gps.hiv}, [], [], [], 0);
[i, s, d, lim] = get_addresses({states2, gps.age, gps.born, gps.strains, gps.hiv}, i, s, d, lim);
d = char(d);

%s.migr       = [s.migr_rect, s.migr_long];
s.allI       = [s.I, s.I2];
% s.migrstates = [i.U.migr_rect, i.Lf.migr_rect, i.Ls.migr_rect, i.Pf.migr_rect, i.Ps.migr_rect];
%s.migrstates = intersect([s.U, s.Lf, s.Ls, s.Pf, s.Ps],s.migr_rect);
s.infectious = [s.allI, (s.Tx)];
s.prev       = [s.allI, (s.Tx)];

% Include the auxiliaries
names = {'inc','incsources','mort','nTPT', 'ch_notifs'};
lgths = [    6,          36,     1,     1,           1];
for ii = 1:length(names)
    inds = lim + [1:lgths(ii)];
    i.aux.(names{ii}) = inds;
    lim = inds(end);
end
i.nx = lim;


% -------------------------------------------------------------------------
% --- Set up selectors and aggregators

% --- Incidence
tmp = zeros(6,i.nstates); 
tmp(1,s.allI) = 1;
tmp(2,intersect(s.allI,s.vuln)) = 1;
tmp(3,intersect(s.allI,s.ch))   = 1;
tmp(4,intersect(s.allI,s.neg))   = 1;
tmp(5,intersect(s.allI,s.pos))   = 1;
tmp(6,intersect(s.allI,s.art))   = 1;
agg.inc = sparse(tmp);

tmp = zeros(i.nstates);
tmp(s.allI,:) = 1;
% Remove transitions due to change in vulnerability status
tmp(s.dom, s.vuln) = 0;
tmp(s.vuln, s.dom) = 0;
% Remove transitions due to change in age status
tmp(s.ch, s.ad) = 0;
tmp(s.ad, s.ch) = 0;
% Remove transitions due to change in HIV status
tmp(s.pos, s.neg) = 0;
tmp(s.neg, s.pos) = 0;
tmp(s.art,s.pos) = 0; 
tmp(s.pos,s.art) = 0;
sel.inc = tmp - diag(diag(tmp));


% --- Sources of incidence 

set1 = {s.dom, s.vuln};
% set1 = {s.dom, s.migr};
set2 = {s.ds};

tmp  = zeros(length(set1)*length(set2),i.nstates);
row  = 1;
for is1 = 1:length(set1)
    for is2 = 1:length(set2)
        tmp(row, intersect(s.allI, intersect(set1{is1}, set2{is2}))) = 1;
        row = row+1;
    end
end
agg.incsources = sparse(tmp);


% --- Selectors for different origins 

% Untreated TB infection
tmp = zeros(i.nstates);
tmp(s.allI, [s.Lf, s.Ls]) = 1;
sel.L2I = sparse(tmp - diag(diag(tmp)));

% Treated TB infecction
tmp = zeros(i.nstates);
tmp(s.allI, [s.Pf, s.Ps]) = 1;
sel.P2I = sparse(tmp - diag(diag(tmp)));

% Relapse, non-post-treatment
tmp = zeros(i.nstates);
tmp(s.allI, [s.R, s.Rhi]) = 1;
sel.R2I = sparse(tmp - diag(diag(tmp)));

% Relapse, post-treatment
tmp = zeros(i.nstates);
tmp(s.allI, s.Rlo) = 1;
sel.T2I = sparse(tmp - diag(diag(tmp)));


% --- People starting TPT
tmp = zeros(i.nstates);
tmp([s.Pf, s.Ps],[s.Lf, s.Ls]) = 1;
sel.nTPT = tmp - diag(diag(tmp));

% --- Notifications
tmp = zeros(i.nstates);
tmp(intersect([s.Tx, s.Tx2],s.ch),:) = 1;
% Remove transitions due to change in vulnerability status
tmp(s.dom, s.vuln) = 0;
tmp(s.vuln, s.dom) = 0;
% Remove transitions due to change in age status
tmp(s.ch, s.ad) = 0;
tmp(s.ad, s.ch) = 0;
% Remove transitions due to change in HIV status
tmp(s.pos, s.neg) = 0;
tmp(s.neg, s.pos) = 0;
tmp(s.art,s.pos) = 0; 
tmp(s.pos,s.art) = 0;
sel.ch_notifs = tmp - diag(diag(tmp));


% -- Natural history parameters -------------------------------------------
r.progression0  = 0.0826;
r.LTBI_stabil   = 0.872;
r.reactivation0 = 0.0006;

r.Tx            = 2;
p.RRrec         = 1;
r.RR_acqu       = 0;
r.Tx2           = 9/12;
r.ltfu          = 0.01;                                                     % loss to followup
r.ltfu2         = r.Tx2*2;

r.self_cure    = 1/6;
r.relapse      = [0.032 0.14 0.0015];
% r.relapse      = [0 0 0];
% r.mu           = 1/66;                                                   % natural mortality
r.muTB         = 1/6;                                                      % TB related mortality
p.imm          = 0.8;                                                      % Reduced susceptibility conferred by previous infection

%p.ch_in_migr   = 0.0789;                                                      % DOUBLE check with country data

% --- Interventions 
p.migrTPT      = 0;                                                        % Proportion of migrants initiated on TPT on entry
% p.TPTeff       = [0.6 0.1];                                                % Effectiveness of TPT
p.TPTeff       = 0.6;                                                % Effectiveness of TPT
r.TPT          = [0 0 0 0];                                                % Uptake of TPT amongst: 1.domestic, 2.recent migrants, 3.long-term migrants
r.TPT2020rec   = 0.004;
r.ACF          = [0 0 0 0];
r.ACF2         = [0 0 0 0];

% p.migrect_popn = 0.437;
% r.migr         = 0.0847;                                                   % https://doi.org/10.1007/s44197-022-00040-w

% p.LTBI_in_migrad = 0.17;
% p.LTBI_in_migrch = 0.03;


% -------------------------------------------------------------------------
% --- Name free parameters ------------------------------------------------

names = {'beta','betadec','gamma','p_relrate_gamma_chvad','r_vuln_sc','relbeta_vuln', 'p_relrate', 'r_ageing_sc','p_relrate_factor', 'contmat_factor', 'r_ART_init','r_HIV_mort','p_HIV_relrate'};      
lgths = [    1,        1,      2,                 1,        1,             1,           2,              1,                  1,                1,            1,              1,              1];

lim = 0; xi = [];
for ii = 1:length(names)
    inds = lim + [1:lgths(ii)];
    xi.(names{ii}) = inds;
    lim = inds(end);
end

% Set their boundaries
bds = [];
bds(xi.beta,:)                   = [0 40];
bds(xi.betadec,:)                = [0 0.15];
bds(xi.gamma,:)                  = repmat([1e-4 10],2,1);
bds(xi.p_relrate_gamma_chvad,:)  = [0 1];
bds(xi.p_relrate,:)              = repmat([1 20],2,1);
bds(xi.r_vuln_sc,:)              = [0 1];
bds(xi.relbeta_vuln,:)           = [0.1 20];
bds(xi.r_ageing_sc,:)            = [0 1];
bds(xi.p_relrate_factor,:) = [1, 10];
bds(xi.contmat_factor,:)    = [0, 1];
bds(xi.r_ART_init,:)    = [0 10];
bds(xi.r_HIV_mort,:)    = [0 10];
bds(xi.p_HIV_relrate,:) = [1 100];
prm.bounds = bds';

ref.i = i; ref.s = s; ref.xi = xi;
prm.p = p; prm.r = r; prm.agg = agg; prm.sel = sel;

prm.contmat_born = [1, 0.5, 0.2; 0.5, 1, 0.2; 0.2 0.2 1];
prm.contmat_age  = [0.2830 0.2525; 0.0692 0.3953];

prm.contmat      = zeros(4, 4);
% go through each element
for age_row = 1:2                                                           % rows in age
    for age_col = 1:2                                                       % cols in age
        for born_row = 1:2                                                  % rows in born
            for born_col = 1:2                                              % cols in born
                % calc position in combined matrix
                row = (born_row-1)*2 + age_row;                             % correctly scale current rows into new matrix
                col = (born_col-1)*2 + age_col;                             % correctly scale current cols into new matrix
                % multiply
                prm.contmat(row, col) = prm.contmat_born(born_row, born_col) * prm.contmat_age(age_row, age_col);
            end
        end
    end
end



load data_Thembisa_AFR.mat;

% -------------------------------------------------------------------------
% --- Specify --------------------------------------------------------

data.incd2010       = [10 12 14];                                           % With broader uncertainty intervals
data.incd2020       = [6.8  7.9  9.2];                                             
data.mort           = [0.26 0.36 0.47];                                     % TB mortality, 2020
data.p_vulnpopn     = [8 10 12]/100;                                        % Proportion of UK population being vulnerable
data.p_vulnTB       = [5 7 9]/100;                                          % Proportion contribution to overall incidence
data.nTPT2019       = 1.3*[0.9 1 1.1];                                      % Number of TPT initiations in 2019, per 10^5 population
data.propincd_ch    = [0.006 0.014 0.025];
data.p_chpopn       = [0.198 0.2471 0.3];                                    % proportion of country thats children
data.p_adpopn       = [0.65 0.7529 0.85];                                  % proportion of country thats adults
data.ch_notifs      = [3 3.69 4.2]/4.576e6*1e5;                             % notifications in the country  
data.HIV_prev       = HIVprev_2019;
data.ART_covg       = ARTcovg_2019;
prm.ART_start       = ART_start;

ys1 = HIV_incd(:,2);
xs1 = 1:length(ys1);
xs2 = 1:length(ys1)+30;
ys2 = max(interp1(xs1,ys1,xs2,'linear','extrap'),0);
prm.rHIV = ys2;

show = 0;
f1a = get_distribution_fns(data.incd2010,   'lognorm', show);
f1b = get_distribution_fns(data.incd2020,   'lognorm', show);
f2  = get_distribution_fns(data.mort,       'lognorm', show);
f3  = get_distribution_fns(data.p_vulnpopn, 'beta',    show);
f4  = get_distribution_fns(data.p_vulnTB,   'beta',    show);
f5  = get_distribution_fns(data.propincd_ch,'beta', show);
f6  = get_distribution_fns(data.p_chpopn,  'beta',    show);
f7  = get_distribution_fns(data.p_adpopn,  'beta',    show);
f8  = get_distribution_fns(data.ch_notifs, 'lognorm', show);
f9 = get_distribution_fns(data.ART_covg,    'beta', show);
f10 = get_distribution_fns(data.HIV_prev,    'lognorm', show);


% lhd.fn = @(incd, mort, p_migrTB, p_migrpopn, p_LTBI) f1(incd) + f2(mort) + f3(p_migrTB) + f4(p_migrpopn) + f5(p_LTBI);
% lhd.fn = @(incd2010, incd2020, mort, p_migrTB, p_migrpopn, p_LTBI, nTPT2019) f1a(incd2010) + f1b(incd2020) + f2(mort) + f3(p_migrTB) + f4(p_migrpopn) + f5(p_LTBI) + f6(nTPT2019);
lhd.fn = @(incd2010, incd2020, mort, p_vulnpopn, p_vulnTB, p_chpopn, ch_notifs, ART_covg, HIV_prev) f1a(incd2010) + f1b(incd2020) + ...
                                                                                                    f2(mort) + f3(p_vulnpopn) + f4(p_vulnTB) ...
                                                                                                    + f6(p_chpopn) + f8(ch_notifs) ...
                                                                                                    + f9(ART_covg) + f10(HIV_prev);

save Model_setup;

File: Python/setup_model.py
from get_addresses import get_addresses
import numpy as np
from scipy import sparse
from get_distribution_fns import get_distribution_fns
import pickle

states1 = ['U']
states2 = ['Lf', 'Ls', 'Pf', 'Ps', 'I', 'I2', 'Tx', 'Tx2', 'Rlo', 'Rhi', 'R']

gps = {
    'age': ['ch', 'ad'],
    'born': ['dom'],
    'hiv': ['neg', 'pos', 'art']
}

i, s, d, lim = get_addresses([states1, gps['age'], gps['born'], gps['hiv']], {}, {}, {}, 0)
i, s, d, lim = get_addresses([states2, gps['age'], gps['born'], gps['hiv']], i, s, d, lim)
d = {k: ' '.join(v) for k, v in d.items()}

s['allI'] = s['I'] + s['I2']
s['infectious'] = s['allI'] + s['Tx']
s['prev'] = s['allI'] + s['Tx']

# Include the auxiliaries
names = ['inc', 'incsources', 'mort', 'nTPT', 'ch_notifs']
lgths = [5, 14, 1, 1, 1]
for name, lg in zip(names, lgths):
    inds = list(range(lim + 1, lim + lg + 1))
    if 'aux' not in i:
        i['aux'] = {}
    i['aux'][name] = inds
    lim = inds[-1]
i['nx'] = lim


# -------------Set up selectors and aggregators--------------------------------

# ~~~~~~~~~Incidence
tmp = np.zeros((5, i['nstates']))
tmp[0, np.array(s['allI']) - 1] = 1
tmp[1, np.array(np.intersect1d(s['allI'], s['ch'])) - 1] = 1
tmp[2, np.array(np.intersect1d(s['allI'], s['neg'])) - 1] = 1
tmp[3, np.array(np.intersect1d(s['allI'], s['pos'])) - 1] = 1
tmp[4, np.array(np.intersect1d(s['allI'], s['art'])) - 1] = 1
agg = {}
agg['inc'] = sparse.csr_matrix(tmp)

tmp = np.zeros((i['nstates'], i['nstates']))
tmp[np.array(s['allI']) - 1, :] = 1
tmp[np.ix_(np.array(s['ch']) - 1, np.array(s['ad']) - 1)] = 0
tmp[np.ix_(np.array(s['ad']) - 1, np.array(s['ch']) - 1)] = 0
tmp[np.ix_(np.array(s['pos']) - 1, np.array(s['neg']) - 1)] = 0
tmp[np.ix_(np.array(s['neg']) - 1, np.array(s['pos']) - 1)] = 0
tmp[np.ix_(np.array(s['art']) - 1, np.array(s['pos']) - 1)] = 0
tmp[np.ix_(np.array(s['pos']) - 1, np.array(s['art']) - 1)] = 0
sel = {}
sel['inc'] = tmp - np.diag(np.diag(tmp))
sel['inc'] = np.array(sel['inc'], dtype=float)



# -----------Sources of incidence 
set1 = [s['dom']]
tmp = np.zeros((len(set1), i['nstates']))
row = 0
for s_set in set1:
    inds = np.intersect1d(np.array(s['allI']), np.array(s_set))
    tmp[row, (inds - 1)] = 1
    row += 1
agg['incsources'] = sparse.csr_matrix(tmp)


# ----People starting TPT
tmp = np.zeros((i['nstates'], i['nstates']))
row_idx = np.concatenate((np.array(s['Pf']), np.array(s['Ps']))) - 1
col_idx = np.concatenate((np.array(s['Lf']), np.array(s['Ls']))) - 1
tmp[np.ix_(row_idx, col_idx)] = 1
sel['nTPT'] = sparse.csr_matrix(tmp - np.diag(np.diag(tmp)))




# -----Notifications
tmp = np.zeros((i['nstates'], i['nstates']))
sel_indices = np.intersect1d(
    np.concatenate((np.array(s['Tx']), np.array(s['Tx2']))),
    np.array(s['ch'])
) - 1
tmp[sel_indices, :] = 1
tmp[np.ix_(np.array(s['ch']) - 1, np.array(s['ad']) - 1)] = 0
tmp[np.ix_(np.array(s['ad']) - 1, np.array(s['ch']) - 1)] = 0
tmp[np.ix_(np.array(s['pos']) - 1, np.array(s['neg']) - 1)] = 0
tmp[np.ix_(np.array(s['neg']) - 1, np.array(s['pos']) - 1)] = 0
tmp[np.ix_(np.array(s['art']) - 1, np.array(s['pos']) - 1)] = 0
tmp[np.ix_(np.array(s['pos']) - 1, np.array(s['art']) - 1)] = 0
sel['ch_notifs'] = sparse.csr_matrix(tmp - np.diag(np.diag(tmp)))


# ------ Natural history parameters-------------------------------------------
r = {}
p = {}

r['progression0'] = 0.0826
r['LTBI_stabil'] = 0.872
r['reactivation0'] = 0.0006
r['Tx'] = 2
p['RRrec'] = 1
r['RR_acqu'] = 0
r['Tx2'] = 9/12
r['ltfu'] = 0.01
r['ltfu2'] = r['Tx2'] * 2
r['self_cure'] = 1/6
r['relapse'] = [0.032, 0.14, 0.0015]
r['muTB'] = 1/6
p['imm'] = 0.8
p['migrTPT'] = 0
p['TPTeff'] = 0.6
r['TPT'] = [0, 0, 0, 0]
r['TPT2020rec'] = 0.004
r['ACF'] = [0, 0, 0, 0]
r['ACF2'] = [0, 0, 0, 0]


# ---Name free parameters ----------------------------------------------------
names = ['beta', 'betadec', 'gamma', 'p_relrate_gamma_chvad', 'p_relrate', 'r_ageing_sc', 'p_relrate_factor', 'contmat_factor', 'r_ART_init', 'r_HIV_mort', 'p_HIV_relrate']
lgths = [1,         1,          2,                      1,          2,              1,          1,              1   ,                  1,           1,          1]

lim = 0
xi = {}
for name, lg in zip(names, lgths):
    inds = list(range(lim + 1, lim + lg + 1))
    xi[name] = inds
    lim = inds[-1]

bds = np.zeros((lim, 2))
bds[np.array(xi['beta']) - 1, :] = [0, 40]
bds[np.array(xi['betadec']) - 1, :] = [0, 0.15]
for idx in xi['gamma']:
    bds[idx - 1, :] = [1e-4, 10]
for idx in xi['p_relrate_gamma_chvad']:
    bds[idx - 1, :] = [0, 1]
for idx in xi['p_relrate']:
    bds[idx - 1, :] = [1, 20]
bds[np.array(xi['r_ageing_sc']) - 1, :] = [0, 1]
bds[np.array(xi['p_relrate_factor']) - 1, :] = [1, 10]
bds[np.array(xi['contmat_factor']) - 1, :] = [0, 1]
bds[np.array(xi['r_ART_init']) - 1, :] = [0, 10]
bds[np.array(xi['r_HIV_mort']) - 1, :] = [0, 10]
bds[np.array(xi['p_HIV_relrate']) - 1, :] = [1, 100]
prm = {}
prm['bounds'] = bds.T
prm['p'] = p
prm['r'] = r
prm['agg'] = agg
prm['sel'] = sel
ref = {'i': i, 's': s, 'xi': xi}

prm['contmat_born'] = np.array([[1, 0.5, 0.2],
                                [0.5, 1, 0.2],
                                [0.2, 0.2, 1]])
prm['contmat_age'] = np.array([[0.2830, 0.2525],
                               [0.0692, 0.3953]])
prm['contmat'] = np.zeros((4, 4))
for age_row in range(1, 3):
    for age_col in range(1, 3):
        for born_row in range(1, 3):
            for born_col in range(1, 3):
                row = (born_row - 1) * 2 + age_row
                col = (born_col - 1) * 2 + age_col
                prm['contmat'][row - 1, col - 1] = prm['contmat_born'][born_row - 1, born_col - 1] * prm['contmat_age'][age_row - 1, age_col - 1]


#-----------------Specify ---------------------------------------------------
data = {}
data['incd2010'] = np.array([10, 12, 14])
data['incd2020'] = np.array([6.8, 7.9, 9.2])
data['mort'] = np.array([0.26, 0.36, 0.47])
data['nTPT2019'] = 1.3 * np.array([0.9, 1, 1.1])
data['propincd_ch'] = np.array([0.006, 0.014, 0.025])
data['p_chpopn'] = np.array([0.198, 0.2471, 0.3])
data['p_adpopn'] = np.array([0.65, 0.7529, 0.85])
data['ch_notifs'] = np.array([3, 3.69, 4.2]) / 4.576e6 * 1e5
data['HIV_prev'] = np.array([0.0215, 0.0247, 0.0280])
data['ART_covg'] = np.array([0.5252, 0.6347, 0.7699])
prm['ART_start'] = 2007

ys2 = np.array([
    0, 0.0005, 0.0009, 0.0013, 0.0016, 0.0019, 0.0022, 0.0024, 0.0027, 0.0029, 0.0031, 0.0034, 0.0036, 0.0037, 0.0037, 0.0037, 0.0036, 0.0035, 0.0034,
    0.0033, 0.0031, 0.0030, 0.0028, 0.0027, 0.0026, 0.0024, 0.0023, 0.0022, 0.0021, 0.0020, 0.0019, 0.0017, 0.0016, 0.0014, 0.0013, 0.0012, 0.0012, 0.0011,
    0.0010, 0.0009, 0.0008, 0.0008, 0.0007, 0.0006, 0.0005, 0.0005, 0.0004, 0.0003, 0.0002, 0.0002, 0.0001, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
])
prm['rHIV'] = ys2

show = False
f1a = get_distribution_fns(data['incd2010'], 'lognorm', show)
f1b = get_distribution_fns(data['incd2020'], 'lognorm', show)
f2  = get_distribution_fns(data['mort'], 'lognorm', show)
f5  = get_distribution_fns(data['propincd_ch'], 'beta', show)
f6  = get_distribution_fns(data['p_chpopn'], 'beta', show)
f7  = get_distribution_fns(data['p_adpopn'], 'beta', show)
f8  = get_distribution_fns(data['ch_notifs'], 'lognorm', show)
f9  = get_distribution_fns(data['ART_covg'], 'beta', show)
f10 = get_distribution_fns(data['HIV_prev'], 'lognorm', show)


def lhd_fn(incd2010, incd2020, mort, p_chpopn, ch_notifs, ART_covg, HIV_prev):
    return (f1a(incd2010) + f1b(incd2020) + f2(mort) +
            f6(p_chpopn) + f8(ch_notifs) + f9(ART_covg) + f10(HIV_prev))


# Model_setup = {"lhd_fn": lhd_fn, "data": data, "prm": prm, "ref": ref}
# with open("Model_setup.pkl", "wb") as f:
#     pickle.dump(Model_setup, f)
